TEST_APPS/device/nfcapp/nfctestshim.cpp

```diff
@@ -0,0 +1,392 @@
+    }
+    if (protocols.target_t2t) {
+        strcat(strSupported, "t2t,");
+    }
+    if (protocols.target_t3t) {
+        strcat(strSupported, "t3t,");
+    }
+    if (protocols.target_iso_dep) {
+        strcat(strSupported, "isodep,");
+    }
+    if (protocols.target_nfc_dep) {
+        strcat(strSupported, "nfcdep,");
+    }
+    if (protocols.target_t5t) {
+        strcat(strSupported, "t5t,");
+    }
+    if (strlen(strSupported)) {
+        strSupported[strlen(strSupported) - 1] = '\0'; // strip trailing comma
+    }
+    // return data as text to the plugin framework
+    cmd_printf("{{protocols=%s}}", strSupported);
+    set_last_nfc_error(NFC_OK);
+    cmd_ready(CMDLINE_RETCODE_SUCCESS);
+#endif
+}
+
+void NFCTestShim::cmd_configure_rf_protocols(
+    nfc_rf_protocols_bitmask_t protocols)
+{
+#if MBED_CONF_NFCEEPROM
+    cmd_printf("EEPROM cannot set protocol()\r\n");
+    set_last_nfc_error(NFC_ERR_UNSUPPORTED);
+    cmd_ready(CMDLINE_RETCODE_INVALID_PARAMETERS);
+#else
+
+    nfc_err_t err = set_rf_protocols(protocols);
+    set_last_nfc_error(err);
+    if (NFC_OK != err) {
+        cmd_ready(CMDLINE_RETCODE_FAIL);
+    } else {
+        cmd_ready(CMDLINE_RETCODE_SUCCESS);
+    }
+#endif
+}
+
+/** \brief Copy data from the Controller buffer, or if EEPROM will initiate a read of the
+  * eeprom contents which get dumped as a string {{bytes=XX XX XX XX.. }} and parsed by
+  * the framework
+  * \return void An ICETEA error code and NFC error is set asyncronously
+  */
+void NFCTestShim::cmd_read_nfc_contents()
+{
+#if MBED_CONF_NFCEEPROM
+    ((NFCProcessEEPROM *)this)->queue_read_call();
+    trace_printf("NFCTestShim::cmd_read_nfc_contents() exit\r\n");
+
+#else
+    // returns last message "written", since we cannot read
+    print_ndef_message(_ndef_write_buffer, _ndef_write_buffer_used);
+    trace_printf("Controller buffer data size=%d\r\n", _ndef_write_buffer_used);
+    set_last_nfc_error(NFC_OK);
+
+    trace_printf("NFCTestShim::cmd_read_nfc_contents() exit\r\n");
+    cmd_ready(CMDLINE_RETCODE_SUCCESS);
+#endif
+}
+
+void NFCTestShim::cmd_erase()
+{
+#if MBED_CONF_NFCEEPROM
+    ((NFCProcessEEPROM *)this)->queue_erase_call();
+#else
+    trace_printf("Erase (reset) controller msg buffer\r\n");
+    _ndef_write_buffer_used = 0;
+    memset(_ndef_write_buffer, 0, sizeof(_ndef_write_buffer));
+    set_last_nfc_error(NFC_OK);
+    cmd_ready(CMDLINE_RETCODE_SUCCESS);
+#endif
+}
+
+/** \brief Writes a Text T record buffer with really long message - length checks to be done by driver only.
+  * If an NFC controller, no write to the chip happens, we copy the data into a Controller buffer
+  * \param uri This method must free the passed in pointer
+  * \return void An ICETEA error code and NFC error is set asyncronously
+  */
+void NFCTestShim::cmd_write_long(char *text_string)
+{
+    MessageBuilder builder(_ndef_poster_message);
+    strcpy(::long_string, text_string); //max_ndef - header - overheads
```

Comment:
you already have a heap allocated buffer, you can just keep the pointer and free it at the end without having to keep a statically allocated buffer to keep a copy